Models/USBSpectrometer.cs

```diff
@@ -0,0 +1,696 @@
+        }
+
+        public override void disconnect()
+        {
+            udc.ReleaseInterface(acc.GetInterface(0));
+            udc.Close();
+            isConnected = false;
+            logger.info("closed usb device");
+        }
+
+        public async Task<bool> initAsync()
+        {
+            connect();
+            var pages = await readEEPROMAsync();
+            if (pages is null)
+            {
+                logger.error("Spectrometer.initAsync: failed to read EEPROM");
+                return false;
+            }
+
+            logger.debug("Spectrometer.initAsync: parsing EEPROM");
+            if (!eeprom.parse(pages))
+            {
+                logger.error("Spectrometer.initAsync: failed to parse EEPROM");
+                return false;
+            }
+
+            ////////////////////////////////////////////////////////////////////
+            // post-process EEPROM
+            ////////////////////////////////////////////////////////////////////
+
+            logger.debug("Spectrometer.initAsync: post-processing EEPROM");
+            pixels = eeprom.activePixelsHoriz;
+            laserExcitationNM = eeprom.laserExcitationWavelengthNMFloat;
+
+            logger.debug("Spectrometer.initAsync: computing wavecal");
+            wavelengths = Util.generateWavelengths(pixels, eeprom.wavecalCoeffs);
+
+            if (laserExcitationNM > 0)
+                wavenumbers = Util.wavelengthsToWavenumbers(laserExcitationNM, wavelengths);
+            else
+                wavenumbers = null;
+
+            logger.debug("Spectrometer.initAsync: generating pixel axis");
+            generatePixelAxis();
+
+            // set this early so battery and other BLE calls can progress
+            paired = true;
+
+            ////////////////////////////////////////////////////////////////////
+            // finish initializing Spectrometer 
+            ////////////////////////////////////////////////////////////////////
+
+            //raiseConnectionProgress(1);
+
+            logger.debug("Spectrometer.initAsync: finishing spectrometer initialization");
+            pixels = eeprom.activePixelsHoriz;
+
+            await updateBatteryAsync();
+
+            // for now, ignore EEPROM configuration and hardcode
+            // integrationTimeMS = (ushort)(eeprom.startupIntegrationTimeMS > 0 && eeprom.startupIntegrationTimeMS < 5000 ? eeprom.startupIntegrationTimeMS : 400);
+            // gainDb = eeprom.detectorGain;
+            integrationTimeMS = 400;
+            gainDb = 8;
+
+            verticalROIStartLine = eeprom.ROIVertRegionStart[0];
+            verticalROIStopLine = eeprom.ROIVertRegionEnd[0];
+
+            logger.info($"initialized {eeprom.serialNumber} {fullModelName}");
+            logger.info($"  detector: {eeprom.detectorName}");
+            logger.info($"  pixels: {pixels}");
+            logger.info($"  integrationTimeMS: {integrationTimeMS}");
+            logger.info($"  gainDb: {gainDb}");
```

Comment:
So, we're acting as though all USBSpectrometers have gain (dB), but we also explicitly support Hamamatsu CCD (0x1000) and InGaAs (0x2000) models which certainly don't have gain (dB). Long-term, we should consider subclassing USBSpectrometer into CCDUSBSpectrometer, InGaAsUSBSpectrometer, and CMOSUSBSpectrometer (or add a whole bunch of if-then conditional logic to USBSpectrometer).

I'm not treating this as a release blocker, just pointing out a fairly big "to-do" for the backlog.

Reply:
gain uses the same API for all of those models over USB, does it not? I think this level of abstraction is fine whether we're referring to gainDB or whatever unit CCD/InGaAs use for their gain instead. I don't think we'd gain (haha) much in that class split, though you have a better sense of the differences than I do